src/runtime/interface.go

```diff
@@ -5,25 +5,67 @@ package runtime
 // Interfaces are represented as a pair of {typecode, value}, where value can be
 // anything (including non-pointers).
 
-import "unsafe"
+import (
+	"reflect"
+	"unsafe"
+)
 
 type _interface struct {
 	typecode uintptr
 	value    unsafe.Pointer
 }
 
 // Return true iff both interfaces are equal.
-func interfaceEqual(x, y _interface) bool {
-	if x.typecode != y.typecode {
-		// Different dynamic type so always unequal.
+func interfaceEqual(_x, _y interface{}) bool {
+	return shallowEqual(_x, _y)
+}
+
+func shallowEqual(x, y interface{}) bool {
```

Comment:
Changing the argument types to `reflect.Value` might simplify this a bit. Doing this will mean that you never need to call `.Interface()` and you only need to call `reflect.ValueOf` in `interfaceEqual`.